src/transformers/modeling_tf_utils.py

```diff
@@ -236,78 +236,90 @@ def compute_loss(self, labels, logits):
-                symbolic_weights_names = set(
-                    "/".join(symbolic_weight.name.split("/")[2:]) for symbolic_weight in symbolic_weights
-                )
-                missing_layers.extend(list(symbolic_weights_names - saved_weight_names_set))
-                unexpected_layers.extend(list(saved_weight_names_set - symbolic_weights_names))
+        # Retrieve the name of each layer from the H5 file
+        saved_h5_model_layers_name = set(hdf5_format.load_attributes_from_hdf5_group(f, "layer_names"))
+        model_layers_name_value = {}
 
-    return missing_layers, unexpected_layers
+        # Retrieve the name of each layer from the instanciated model
+        # make it a dict that looks like {"layer_name": Layer object}
+        model_layers_name_value = {layer.name: layer for layer in model.layers}
 
+        # Create a set of unique names of the layers that come from the instanciated model
+        model_layers_name = set(model_layers_name_value.keys())
 
-def load_tf_weights(model, resolved_archive_file):
-    """
-    Load the TF weights from a H5 file.
+        # Find the missing layers from the high level list of layers
+        missing_layers = list(model_layers_name - saved_h5_model_layers_name)
 
-    Args:
-        model (:obj:`tf.keras.models.Model`):
-            The model to load the weights into.
-        resolved_archive_file (:obj:`str`):
-            The location of the H5 file.
-    """
-    with h5py.File(resolved_archive_file, "r") as f:
-        saved_layer_names = set(hdf5_format.load_attributes_from_hdf5_group(f, "layer_names"))
+        # Find the unexpected layers from the high level list of layers
+        unexpected_layers = list(saved_h5_model_layers_name - model_layers_name)
+        saved_weight_names_set = set()
+        symbolic_weights_names = set()
         weight_value_tuples = []
 
-        for layer in model.layers:
-            if layer.name in saved_layer_names:
-                g = f[layer.name]
+        # Compute missing and unexpected sub layers
+        # Store the weights in list of tuples that looks like [(weight_object, value_of_weight),...]
+        for layer_name in saved_h5_model_layers_name:
+            # if layer_name from the H5 file belongs to the layers from the instanciated model
+            if layer_name in model_layers_name:
+                # Get layer_name from the H5 file
+                g = f[layer_name]
+                # Get all the weights that are attach to layer_name
                 saved_weight_names = hdf5_format.load_attributes_from_hdf5_group(g, "weight_names")
+                # Get the layer object from the layer_name in the dict that represents the instanciated model
+                layer = model_layers_name_value[layer_name]
+                # Get all the weights as a list from the layer object
                 symbolic_weights = layer.trainable_weights + layer.non_trainable_weights
                 saved_weight_names_values = {}
```

Comment:
it's a nit, but I think it's cleaner to name this dict just `saved_weights`. `saved_weight_names_values` makes me think of a list of values not a key/value dict...

Reply:
Updated!